### ISSUE TITLE
feature request: allow changing the labels of the legends

### ISSUE THREAD

**Issue by silberzwiebel** [stan-dev/bayesplot#75]
This is an idea that came up in the discussion of #73 

The point is to let users change the labels of the legend in bayesplot objects easily, possibly by setting them as arguments in the plot creation. This could for instance be helpful to plot the pp_check for two models and then label `y_rep` as `model A`or `model B`, respectively.
I mainly used `pp_check()` until now but I guess this might be usable for other plotting possibilites as well.

@jgabry wrote in #73:
>  But another option would be to add additional arguments to all of the bayesplot plotting functions so that user's can specify the names to be used in the legend at the same time they create the plot. For example, it could look something like this maybe: 
>
>```
> my_legend <- list(y = "emp. data", yrep = "model predict.") 
> ppc_bars(y, yrep, ..., legend_labels = my_legend) 
> ```
> That way the legends could be created with those labels instead of having to change them after the fact. I'd need to figure out the best way to implement that (it would vary depending on the plotting function) but that should be much easier to do.

**Comment by jgabry** [comment]
Thanks for opening the issue. This is officially on the to-so list now!

**Comment by silberzwiebel** [comment]
Just stumbling over this again. The work-around for changing the text of the legend still works but now I wanted to change the text of facet strings. In my concrete example I want to contrast the outcome for a boolean factor and I'm just seeing "TRUE" and "FALSE"  as facet labels. However, manual context-sensitive strings would be way more intuitive for the plot.
Should I update the the title of this issue to include other labels than the legend, too, or rather open a new feature request?

**Comment by jgabry** [comment]
@silberzwiebel Thanks for following up on this! 

I would much rather add helper functions for this to modify these things after the plot is created (e.g., like these [other ones](http://mc-stan.org/bayesplot/reference/bayesplot-helpers.html) we have for stuff like changing axis titles) than add many more arguments to the plotting functions, but as far as I know it's not possible to change the legend or facet labels after the fact. @tjmahr Any clever ideas in this regard? Do you know of a way to edit the created ggplot object to change this stuff (maybe messing with `ggplot_build`)? 

In other words, if we can allow this by adding a few helper functions it would be so much better than having to add code for this to each of the plotting functions individually.

**Comment by silberzwiebel** [comment]
Helper functions sound great.
What you [wrote the other year for changing the legend strings was](https://github.com/stan-dev/bayesplot/issues/73#issuecomment-282820138):

> p$scales$scales[[1]]$labels <- "emp. data" 
> p$scales$scales[[2]]$labels <- "model predict."

This happens after the plot is already built (and works fine), so it could go into a helper function. But I'm entirely unsure and unexperienced whether these scales are "stable" in the sense that they will always correspond to the legend and whether it is possible to change other strings like the facet labels in this way.

**Comment by jgabry** [comment]
> This happens after the plot is already built (and works fine), so it could go into a helper function. But I'm entirely unsure and unexperienced whether these scales are "stable" in the sense that they will always correspond to the legend and whether it is possible to change other strings like the facet labels in this way.

Yeah, what's tricky about that is the order of `scales` may be different for the different plots. So it would be great to have a function to use like 

```
p + legend_labels(c("empirical data", "model prediction"))
```
but my hunch is that `legend_labels` would have to have lots of different logic inside to deal with the different possible plots. Maybe possible though. 

Regarding the facet labels, can you give more details about what you want to do? It should actually already be possible to change the facet labels for many of the plotting functions, but you have to use a ggplot `labeller`function as described in `?ggplot2::labeller`. For example, here's how you would do it for `ppc_bars_grouped`:

```r
# simulate some data
f <- function(N) {
  sample(1:4, size = N, replace = TRUE, prob = c(0.25, 0.4, 0.1, 0.25))
}
y <- f(100)
yrep <- t(replicate(500, f(100)))
dim(yrep)
group <- gl(2, 50, length = 100, labels = c("GroupA", "GroupB"))

# facet labels default to labels of 'group' (in this case GroupA, GroupB)
p <- ppc_bars_grouped(y, yrep, group, prob = 0.5, freq = FALSE)
plot(p)
```
![bars_grouped](https://user-images.githubusercontent.com/7796803/38744990-525165b2-3f11-11e8-83fa-d51a53e3139c.png)


Now we can change the facet labels to "Banana" and "Turtle" instead of "GroupA" and "GroupB":

```r
my_labels <- as_labeller(c("GroupA" = "Banana", "GroupB" = "Turtle"))
p + facet_wrap("group", labeller = my_labels)
```

![bars_grouped_relabel](https://user-images.githubusercontent.com/7796803/38745061-82f7b162-3f11-11e8-89c6-25fdf1ef128d.png)

**Comment by jgabry** [comment]
Not too ignorant! It’s not currently in any examples anywhere, but it would
be a welcome addition. For this we could also add a `facet_relabel` helper
function.

**Comment by jgabry** [comment]
Right now there’s not one place to put it that makes sense since it applies to any of the ones with facets. It probably makes sense to add a `facet_relabel()` helper function and then put an example in the documentation in 

https://github.com/stan-dev/bayesplot/blob/master/R/bayesplot-helpers.R

How about I do the function and then you can add an example you like to the doc? I could easily add the example, but I like encouraging users to contribute! Maybe you’ll be more likely to suggest another useful addition in the future ;)

**Comment by silberzwiebel** [comment]
Sounds great!

**Comment by jgabry** [comment]
@silberzwiebel I just pushed a branch called `facet_relabel` that has new functions `facet_relabel_gg` (do the facet relabeling) and `facet_vars` (get info about variable and scales bayesplot used for faceting). There's documentation explaining them too that you can look at for the details. 

I added one example using `mcmc_trace`, but it would great if you can add one for PPC grouped plots too. I made a spot for your example here: 

https://github.com/stan-dev/bayesplot/blob/6b86e130d30be773f73d7527a1def46b84800c41/R/bayesplot-helpers.R#L236

To add the example you can fork the bayesplot repo, switch to the `facet_relabel` branch, add the  example, and then submit a PR to the `facet_relabel` branch (not `master`). Then I'll merge your PR and make a PR against master to get it into the next release. Let me know if you have any questions about that process and thanks for your willingness to help!

**Comment by jgabry** [comment]
Hah, thanks @tjmahr! Seems obvious in hindsight (especially since I used
that method inside of bayesplot) but totally slipped my mind.

**Comment by silberzwiebel** [comment]
@jgabry just to let you know that this is still on my list, I' just busy doing other stuff right now (as always ;)).

**Comment by jgabry** [comment]
Thanks, no rush!

**Comment by jgabry** [comment]
@silberzwiebel I was going through open issues and just noticed I hadn't merged your PR (I could have sworn I did that months ago!). So I just finally merged your PR into my `facet_relabel` branch (not master branch yet). Sorry that took so long. I will now make a PR to master from `facet_relabel` so @tjmahr can take a look and see if we should get this into a release or not.

**Comment by silberzwiebel** [comment]
Thanks! Actually, I was the one who delayed this by months. I opened the PR just at the end of August, so you did all fine.
